### PR TITLE
fixes #12962 - Add Xenserver to list of supported OS families.

### PR DESCRIPTION
**Pull request by kflavin**
Add XenServer OS family support to Foreman.  XenServer requires some special handling, because its PXE config uses four files, and they are spread across different directories on the install media.

This is an extension of @mmoll 's earlier work: https://github.com/mmoll/foreman/commit/682c45ce72c132eb5f3f6dd752d06a7300c2a516

### CODE REVIEW

**Comment by dLobatog** [comment]
[test]

**Comment by dLobatog** [comment]
@kflavin Shouldn't this belong on [foreman-xen](https://github.com/theforeman/foreman-xen/)?

**Comment by domcleal** [comment]
I'm not sure that there are enough extension points in Foreman core to add an operating system family reliably from a plugin yet.

**Comment by mmoll** [comment]
I have yet to test this and also this would need templates like the ones in https://github.com/mmoll/community-templates/commits/xenserver (last two commits). Back then I thought of XenServer as a PXE installable OS like every other, so there's no connection to foreman_xen.

**Comment by kflavin** [comment]
I'll get some templates in there.

**Comment by mmoll** [comment]
@kflavin I think the templates from my community-templates branch referenced above are quite complete, you might need to adapt it to your core PR

**Comment by mmoll** [comment]
also, I think `mboot.c32` should be moved by the installer from the system installed syslinux packages and just assumed to be there

**Comment by kflavin** [comment]
@mmoll I moved over the Xen templates from community-templates.  I also removed mboot.c32.  The default PXE template assumes it to be in the TFTP root directory now.

File: app/models/operatingsystems/xenserver.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
+class Xenserver < Operatingsystem
+  #PXEFILES = {:kernel => "vmlinuz", :initrd => "install.img", :xen => "xen.gz"}
```
Comment (mmoll) [inline app/models/operatingsystems/xenserver.rb]: leftover line

**Comment by mmoll** [comment]
@kflavin Firstly, thanks for following up on this! :bow:

See theforeman/community-templates#250 for an updated set of templates. The templates for DB seeding are generated by `script/sync_templates.sh` after the templates have been pulled in and should not be included in this PR to Foreman core.

In app/models/operatingsystems/xenserver.rb I edit mediumpath to:

``` ruby
  def mediumpath(host)
    medium_uri(host).to_s
  end
```

and in lib/foreman/renderer.rb:

``` ruby
    def xenserver_attributes
      @mediapath = @host.operatingsystem.mediumpath @host
      @xen       = @host.operatingsystem.xen @host.arch
    end
```

for everything to work. After a core dev reviewed the next iteration (and the comments are addressed). I'll test this again.

**Comment by mmoll** [comment]
If anybody else wants to test this: http://xenserver.org/ -> Download and loopback-mount the ISO to some dir that's served by a webserver and use that URL as installation medium.

**Comment by kflavin** [comment]
@mmoll No problem!  I'm excited to have a chance to contribute.

I'll make those adjustments.  How about db/seeds.d/07-provisioning_templates.rb?  Should I revert my changes there as well?

**Comment by mmoll** [comment]
@kflavin yes please. If anybody needs to test this, foreman_templates is able to pull the templates in from the templates PR by having the correct repo/remote specification

**Comment by mmoll** [comment]
also, please rebase

**Comment by kflavin** [comment]
@mmoll I've removed the templates and rebased against the develop branch.

**Comment by mmoll** [comment]
[test]

**Comment by mmoll** [comment]
@kflavin tests fail, the OS list needs updating.

**Comment by mmoll** [comment]
works for me, otherwise...

**Comment by kflavin** [comment]
[test]

**Comment by mmoll** [comment]
[test] <- you need to be in the theforeman organization on Github for that to work

**Comment by kflavin** [comment]
@mmoll I added Xenserver under  test/unit/operatingsystem_test.rb and amended my last commit.  It doesn't look like I have permissions to kick off the tests on jenkins.  They seem to be running locally  now though (I get some other errors which appear to be related phantomjs and my setup).

**Comment by kflavin** [comment]
@mmoll gotcha, thanks :)

**Comment by mmoll** [comment]
@kflavin please check and fix the codestyle (rubocop) notices, the one failure in one testset looks unrelated

File: app/models/operatingsystems/xenserver.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+  def self.model_name
+    superclass.model_name
+  end
+
+  def bootfile(arch, type)
+    pxe_prefix(arch) + "-" + eval("#{self.family}::PXEFILES[:#{type}]").split("/")[-1]
+  end
+
+  def boot_files_uri(medium, architecture, host = nil)
```
Comment (mmoll) [inline app/models/operatingsystems/xenserver.rb:37]: Could you add the URI of mboot.c32 also? I didn't try if it works, but that would open up the possibility to have all needed URLs for an iPXE template.

**Comment by kflavin** [comment]
@mmoll Ah, didn't realize what the rubocop tests were.  I fixed up the code style changes it complained about.  I didn't get to your second comment yet, but I'll come back and address that as well.

**Comment by mmoll** [comment]
two hints on that:
- look at the iPXE templates for preseed and autoyast.
- check the following rails console snippet:

```
irb(main):001:0> h = Host.find_by_name('xenserver.mystuff.org')
[...]
irb(main):002:0> boot_files_uris = h.operatingsystem.boot_files_uri(h.medium,h.architecture)
=> [#<URI::HTTP:0x0000000b632120 URL:http://webserver.mystuff.org/xs/6.5//boot/vmlinuz>, #<URI::HTTP:0x0000000b6310e0 URL:http://webserver.mystuff.org/xs/6.5//install.img>, #<URI::HTTP:0x0000000b630500 URL:http://webserver.mystuff.org/xs/6.5//boot/xen.gz>]
```

iPXE would also need mboot.c32 there (as said, I didn't test if it really works via iPXE, but that would be nice)

**Comment by domcleal** [comment]
[test]

**Comment by kflavin** [comment]
@mmoll I probably should have made a new commit instead of amending the last one, so you could see the changes.

Here's what I modified in app/models/operatingsystems/xenserver.rb.  One line added, one modified:

```
+ MBOOT = "boot/pxelinux/mboot.c32"

-eval("#{self.family}::PXEFILES").values.collect do |img|
+eval("#{self.family}::PXEFILES").values.push(MBOOT).collect do |img|
```

And here is is what is returned in the console:

```
irb(main):001:0> h = Host.find_by_name('myxenserver.domain')
irb(main):010:0> h.operatingsystem.pxe_files(h.medium, h.architecture)
=> [{:"boot/Xenserver-6.5-x86_64"=>"http://myserver/mrepo/XenServer-6.5-x86_64/disc1/boot/vmlinuz"}, {:"boot/Xenserver-6.5-x86_64"=>"http://myserver/mrepo/XenServer-6.5-x86_64/disc1/install.img"}, {:"boot/Xenserver-6.5-x86_64"=>"http://myserver/mrepo/XenServer-6.5-x86_64/disc1/boot/xen.gz"}, {:"boot/Xenserver-6.5-x86_64"=>"http://myserver/mrepo/XenServer-6.5-x86_64/disc1/boot/pxelinux/mboot.c32"}]
```

**Comment by mmoll** [comment]
[test]
I did craft an iPXE template, however after loading all the files via HTTP it just doesn't boot, just hanging, anyway I'll leave that out of the templates PR.

**Comment by kflavin** [comment]
@mmoll Anything else I should do here?

**Comment by mmoll** [comment]
@kflavin nope, all fine :+1:
we now need to wait for somebody with push permissions to Foreman core to review

**Comment by domcleal** [comment]
Merged as 7cac089 for Foreman 1.12.  Thanks for the patch @kflavin!

**Pull request closed without merging**